src/test/csf/Scheduler.h

```diff
@@ -0,0 +1,497 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012-2017 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_CSF_SCHEDULER_H_INCLUDED
+#define RIPPLE_TEST_CSF_SCHEDULER_H_INCLUDED
+
+#include <ripple/basics/qalloc.h>
+#include <ripple/beast/clock/manual_clock.h>
+#include <boost/intrusive/set.hpp>
+#include <type_traits>
+#include <utility>
+
+namespace ripple {
+namespace test {
+namespace csf {
+
+/** Simulated discrete-event scheduler.
+
+    Simulates the behavior of events using a single common clock.
+
+    An event is modeled using a lambda function and is scheduled to occur at a
+    specific time. Events may be canceled using a token returned when the
+    event is scheduled.
+
+    The caller uses one or more of the step, step_one, step_for, step_until and
+    step_while functions to process scheduled events.
+
+    This class is meant to be used in conjunction with BasicNetwork to allow
+    simulating event processing and messaging.
+*/
+class Scheduler
+{
+public:
+    using clock_type = beast::manual_clock<std::chrono::steady_clock>;
+
+    using duration = typename clock_type::duration;
+
+    using time_point = typename clock_type::time_point;
+
+private:
+    // Allow BasicNetwork the ability to schedule events with drop conditions
+    template <class T>
+    friend class BasicNetwork;
```

Comment:
Agreed.  I'm actually in the process of an alternate design and will probably just close this PR.